doc/timelineplot.1
.\"
.\" Documentation for timelineplot
.\" groff/troff/nroff format for linux man page
.\"
.TH TIMELINEPLOT 1 "2019-09-16" "" ""

.SH NAME
timelineplot - generate simple line plots with time on the x-axis


.SH SYNOPSIS
\fBtimelineplot\fR {\fIdatafile\fR|\-} \fIimagefile\fR [\fIoption\fR]...


.SH DESCRIPTION
Generate simple line plots from two columns of character-delimited 
data. The first column represents the time variable (x-axis) and the 
second column contains the values to be plotted.

\fBTimelineplot\fR is a wrapper for \fBgnuplot\fR with many options 
in the plot script adjustable by optional command line arguments.



.SH REQUIRED ARGUMENTS

.TP
\fIdatafile\fR

Two-column text file with time series data to be plotted. The default
column separator is the pipe character, "|". The first column must be 
a date string (default format is "%Y-%m-%d %H:%M:%S"). The second 
column must be numeric. Any additional columns will be ignored. 

Command line options may be used to specify an alternative column
separator or time format string. 

When \fIdatafile\fR is "-", read data from standard input instead of a file.

.TP
\fIimagefile\fR

File to use for the plot output image. This file will be overwritten 
if it already exists.

The output image will be created in \fBpng\fR format by default.



.SH OPTIONS

.SS Input modifiers

.TP
\fB\-sep\fR \fIcharacter\fR

Column separator for the input data file. Default is "|".
(The separator character should be enclosed in double quotes.)

.TP
\fB\-tfmt\fR \fIformat_string\fR

Format string describing the time format in the first column
of the input file. Default value is "%Y-%m-%d %H:%M:%S".


.SS Output modifiers

.TP
\fB\-xfmt\fR \fIformat_string\fR

Format string to use for x-axis labels in the output plot.
Default value is "" (no labels).


.TP
\fB\-h\fR \fIpixels\fR

Height of output image in pixels. Default is 300.


.TP
\fB\-w\fR \fIpixels\fR

Width of output image in pixels. Default is 400.


.TP
\fB\-format\fR \fIformat\fR

Image format for the plot image. Default is "png". Other
choices include "svg", "jpeg", and "gif".


.TP
\fB\-title\fR \fIstring\fR

Title to associate with the y-axis data. Default value is "".


.TP
\fB\-linetitle\fR \fIstring\fR

Title to associate with the plotted line in a key. (Not terribly
useful since we're only plotting one line.) Default value is "".


.TP
\fB\-xlabel\fR \fIstring\fR

Label for x-axis. Default value is "".


.TP
\fB\-ylabel\fR \fIstring\fR

Label for y-axis. Default value is "".


.TP
\fB\-xmin\fR \fItime_expression\fR, \fB\-xmax\fR \fItime_expression\fR

Minimum and maximum values to use for x-axis. Useful if a constant
scale is needed. Default is for the plot to fit the data values.
The time expression must be in the same format as the time values
in the input data.


.TP
\fB\-ymin\fR \fInumber\fR, \fB\-ymax\fR \fInumber\fR

Minimum and maximum values to use for y-axis. Useful if a constant
scale is needed. Default is for the plot to fit the data values.


.TP
\fB\-xrot\fR \fIdegrees\fR

Rotation angle for the timestamp labels on the x-axis tics. 
Default is 60.


.TP
\fB\-lc\fR \fIcolor_expression\fR

Set the color of the line. Default value is "#0000FF".
Any value that gnuplot will accept after "rgbcolor" can be used
here (e.g. "red", "0xff0000", or "#ff0000").


.TP
\fB\-lw\fR \fIpixels\fR

Set the line width in pixels. Default is 2.


.TP
\fB\-bgcolor\fR \fIcolor_expression\fR

Set the color of the plot background. Default is "#FFFFFF".


.TP
\fB\-grid\fR {xtics|ytics|"noxtics noytics"}

Set the background grid. Default is to show the full grid.
.br
\fBxtics\fR - only show gridlines parallel to x-axis
.br
\fBytics\fR - only show gridlines parallel to y-axis
.br
\fB"noxtics noytics"\fR - no background grid


.TP
\fB\-gridcolor\fR \fIcolor_expression\fR

Set the color of background grid lines. Default is "#B0B0B0".


.TP
\fB\-font\fR \fIfontname\fR

Specify the font. Default is "arial".


.TP
\fB\-fontsize\fR \fInumber\fR

Specify the font size. Default is 12.


.TP
\fB\-fontcolor\fR \fIcolor_expression\fR

Set the color of the fonts and borders. Default is "#000000".



.SH AUTHOR
Written by Evan Linde (elinde@okstate.edu)



.SH SEE ALSO
gnuplot(1)

More time format strings appear to work than are indicated by gnuplot's 
documentation. See strftime(3) for additional time formatters that might 
work.


timelineplot.sh
#!/bin/bash
#
# Make a simple line plot from a two-column file of time series data.
#
# Usage:
#
#   timelineplot.sh <datafile> <plotout> [options]
#
#     <datafile>
#
#        Two-colum character delimited (default is "|") text file with 
#        time series data to be plotted. First column must be a date
#        parsable by strftime (default format is "%Y-%m-%d %H:%M:%S"). 
#        Second column must be numeric. The separator and time format
#        can be set to different values with command line options.
#        Use "-" to read data from a pipe.
#
#     <plotout>
#
#        File to use for the plot output (image). This file will be
#        overwritten if it already exists.
#
#
#   Optional arguments
#
#   For input
#
#     -sep <character>
#
#        Column separator for the input data file. Default is "|".
#        (The separator character should be enclosed in double quotes.)
#
#     -tfmt <format_string>
#
#        Format string describing the time format in the first column
#        of the input file. Default value is "%Y-%m-%d %H:%M:%S".
#
#   For output
#
#     -xfmt <format_string>
#
#        Format string to use for x-axis labels in the output plot.
#        Default value is "" (no labels).
#
#     -h <pixels>
#
#        Height of output image in pixels. Default is 300.
#
#     -w <pixels>
#
#        Width of output image in pixels. Default is 400.
#
#     -format <format>
#
#        Image format for the plot image. Default is "png". Other
#        choices include "svg", "jpeg", and "gif".
#
#     -title <string>
#
#        Title to associate with the y-axis data. Default value is "".
#
#     -linetitle <string>
#
#        Title to associate with the plotted line in a key. (Not terribly
#        useful since we're only plotting one line.) Default value is "".
#
#     -xlabel <string>
#
#        Label for x-axis. Default value is "".
#
#     -ylabel <string>
#
#        Label for y-axis. Default value is "".
#
#     -xmin <time_expression>
#     -xmax <time_expression>
#
#        Minimum and maximum values to use for x-axis. Useful if a constant
#        scale is needed. Default is for the plot to fit the data values.
#        The time expression must be in the same format as the time values
#        in the input data.
#
#     -ymin <number>
#     -ymax <number>
#
#        Minimum and maximum values to use for y-axis. Useful if a constant
#        scale is needed. Default is for the plot to fit the data values.
#
#     -xrot <degrees>
#
#        Rotation angle for the timestamp labels on the x-axis tics. 
#        Default is 60.
#
#     -lc <color_expression>
#
#        Set the color of the line. Default value is "#0000FF".
#        Any value that gnuplot will accept after "rgbcolor" can be used
#        here (e.g. "red", "0xff0000", or "#ff0000").
#
#     -lw <pixels>
#
#        Set the line width in pixels. Default is 2.
#
#     -bgcolor <color_expression>
#
#        Set the color of the plot background. Default is "#FFFFFF".
#
#     -grid <"" | "xtics" | "ytics" | "noxtics noytics">
#
#        Set the background grid. Default is "".
#        "" - show full grid
#        "xtics" - only show gridlines parallel to x-axis
#        "ytics" - only show gridlines parallel to y-axis
#        "noxtics noytics" - no background grid
#
#     -gridcolor <color_expression>
#
#        Set the color of background grid lines. Default is "#B0B0B0".
#
#     -font <fontname>
#
#        Specify the font. Default is "arial".
#
#     -fontsize <number>
#
#        Specify the font size. Default is 12.
#
#     -fontcolor <color_expression>
#
#        Set the color of the fonts and borders. Default is "#000000".
#  
#

err=0
datafile="$1"
plotout="$2"
shift 2
while (( $# )); do
    case $1 in
      -sep) sep="$2"; shift 2
          ;;
      -tfmt) timefmt="$2"; shift 2
          ;;
      -xfmt) xtimefmt="$2"; shift 2
          ;;
      -h) height="$2"; shift 2
          ;;
      -w) width="$2"; shift 2
          ;;
      -format) format="$2"; shift 2
          ;;
      -title) title="$2"; shift 2
          ;;
      -linetitle) linetitle="$2"; shift 2
          ;;
      -xlabel) xlabel="$2"; shift 2
          ;;
      -ylabel) ylabel="$2"; shift 2
          ;;
      -xmin) xmin="$2"; shift 2
          ;;
      -xmax) xmax="$2"; shift 2
          ;;
      -ymin) ymin="$2"; shift 2
          ;;
      -ymax) ymax="$2"; shift 2
          ;;
      -xrot) xrot="$2"; shift 2
          ;;
      -lc) linecolor="$2"; shift 2
          ;;
      -lw) linewidth="$2"; shift 2
          ;;
      -bgcolor) bgcolor="$2"; shift 2
          ;;
      -grid) grid="$2"; shift 2
          ;;
      -gridcolor) gridcolor="$2"; shift 2
          ;;
      -font) font="$2"; shift 2
          ;;
      -fontsize) fontsize="$2"; shift 2
          ;;
      -fontcolor) fontcolor="$2"; shift 2
          ;;
      *) echo "Unexpected parameter: $1" 1>&2; ((err++)); shift
          ;;
    esac
done

# Set any unset variables to default values 
: ${sep:="|"}
: ${timefmt:="%Y-%m-%d %H:%M:%S"}
: ${xtimefmt:=""}
: ${height:=300}
: ${width:=400}
: ${format:="png"}
: ${title:=""}
: ${linetitle:=""}
: ${xlabel:=""}
: ${ylabel:=""}
: ${xmin:=""}
: ${xmax:=""}
: ${ymin:=""}
: ${ymax:=""}
: ${xrot:=60}
: ${linecolor:="#0000FF"}
: ${linewidth:=2}
: ${bgcolor:="#FFFFFF"}
: ${grid:=""}
: ${gridcolor:="#B0B0B0"}
: ${font:="arial"}
: ${fontsize:=12}
: ${fontcolor:="#000000"}

# Add quotes around xmin and xmax if they're non-empty; need to remain unquoted if empty
[[ ${#xmin} -gt 0 ]] && xmin="\"${xmin}\""
[[ ${#xmax} -gt 0 ]] && xmax="\"${xmax}\""

xtics_align="right"
[[ ${xrot} -eq 0 ]] && xtics_align="center"

# Handle piped input by creating a temp file
if [[ "${datafile}" == "-" ]]; then
    T=$(mktemp) 
    cat > $T
    datafile="$T"
fi

gnuplot << EOF > "${plotout}"
set datafile separator "${sep}"
set terminal ${format} size ${width},${height} font "${font},${fontsize}" background "${bgcolor}"
set title "${title}" tc "${fontcolor}"
set yrange [${ymin}:${ymax}]
set ylabel "${ylabel}" tc "${fontcolor}"
set xlabel "${xlabel}" tc "${fontcolor}"
set xdata time
set timefmt "${timefmt}" 
set xrange [${xmin}:${xmax}]
set format x "${xtimefmt}"
set xtics ${xtics_align} rotate by ${xrot} font "${font},${fontsize}" tc "${fontcolor}"
set ytics tc "${fontcolor}"
set border lc "${fontcolor}"
set key left top tc "${fontcolor}"
set grid ${grid} lc "${gridcolor}" # don't quote ${grid} since gnuplot won't like the empty quotes
plot "${datafile}" using 1:2 with lines lw ${linewidth} lt 1 lc rgbcolor "${linecolor}" title "${linetitle}"
EOF

# Delete the temp file if we created one
[[ -f "$T" ]] && rm -f "$T"
